File: ch4/kruskal_prim.java
import java.util.*;
import java.io.*;

// Union-Find Disjoint Sets Library written in OOP manner, using both path compression and union by rank heuristics
class UnionFind {                                              // OOP style
  private ArrayList<Integer> p, rank, setSize;
  private int numSets;

  public UnionFind(int N) {
    p = new ArrayList<Integer>(N);
    rank = new ArrayList<Integer>(N);
    setSize = new ArrayList<Integer>(N);
    numSets = N;
    for (int i = 0; i < N; i++) {
      p.add(i);
      rank.add(0);
      setSize.add(1);
    }
  }

  public int findSet(int i) { 
    if (p.get(i) == i) return i;
    else {
      int ret = findSet(p.get(i)); p.set(i, ret);
      return ret; } }

  public Boolean isSameSet(int i, int j) { return findSet(i) == findSet(j); }

  public void unionSet(int i, int j) { 
    if (!isSameSet(i, j)) { numSets--; 
    int x = findSet(i), y = findSet(j);
    // rank is used to keep the tree short
    if (rank.get(x) > rank.get(y)) { p.set(y, x); setSize.set(x, setSize.get(x) + setSize.get(y)); }
    else                           { p.set(x, y); setSize.set(y, setSize.get(y) + setSize.get(x));
                                     if (rank.get(x) == rank.get(y)) rank.set(y, rank.get(y) + 1); } } }
  public int numDisjointSets() { return numSets; }
  public int sizeOfSet(int i) { return setSize.get(findSet(i)); }
}

public class kruskal_prim {
  static ArrayList<ArrayList<IntegerPair>> AL = new ArrayList<>();
  static ArrayList<Boolean> taken = new ArrayList<>(); // global boolean flag to avoid cycle
  static PriorityQueue<IntegerPair> pq = new PriorityQueue<>(); // priority queue to help choose shorter edges

  static void process(int u) { //  we do not need to use -ve sign to reverse the sort order
    taken.set(u, true);
    for (IntegerPair v_w : AL.get(u))
      if (!taken.get(v_w.first()))
        pq.offer(new IntegerPair(v_w.second(), v_w.first()));
  }

  public static void main(String[] args) throws Exception {
    /*
    // Graph in Figure 4.10 left, format: list of weighted edges
    // This example shows another form of reading graph input
    5 7
    0 1 4
    0 2 4
    0 3 6
    0 4 6
    1 2 2
    2 3 8
    3 4 9
    */

    File f = new File("kruskal_prim_in.txt");
    Scanner sc = new Scanner(f);

    int V = sc.nextInt();
    int E = sc.nextInt();
    // Kruskal's algorithm merged with Prim's algorithm

    AL.clear();
    for (int i = 0; i < V; i++) {
      ArrayList<IntegerPair> Neighbor = new ArrayList<>(); // create ArrayList of pair<int, int> 
      AL.add(Neighbor); // store blank ArrayList first
    }
    ArrayList<IntegerTriple> EL = new ArrayList<>();
  
    // sort by edge weight O(E log E)
    // PQ default: sort descending. Trick: use <(negative) weight(i, j), <i, j> >
    for (int i = 0; i < E; i++) {
      int u = sc.nextInt();
      int v = sc.nextInt();
      int w = sc.nextInt();
      EL.add(new IntegerTriple(w, u, v));                      // (w, u, v)
      AL.get(u).add(new IntegerPair(v, w));
      AL.get(v).add(new IntegerPair(u, w));
    }
    Collections.sort(EL);

    int mst_cost = 0;           // all V are disjoint sets at the beginning
    UnionFind UF = new UnionFind(V);
    for (int i = 0; i < E; i++) {                    // for each edge, O(E)
      IntegerTriple front = EL.get(i);
      if (!UF.isSameSet(front.second(), front.third())) {          // check
        mst_cost += front.first();            // add the weight of e to MST
        UF.unionSet(front.second(), front.third());            // link them
    } }

    // note: the number of disjoint sets must eventually be 1 for a valid MST
    System.out.printf("MST cost = %d (Kruskal's)\n", mst_cost);



  // inside int main() --- assume the graph is stored in AL, pq is empty
    for (int i = 0; i < V; i++)
      taken.add(false);                // no vertex is taken at the beginning
    process(0);   // take vertex 0 and process all edges incident to vertex 0
    mst_cost = 0;
    while (!pq.isEmpty()) { // repeat until V vertices (E=V-1 edges) are taken
      IntegerPair front = pq.peek(); pq.poll();
      int u = front.second(), w = front.first();   // no need to negate id/weight
      if (!taken.get(u)) {           // we have not connected this vertex yet
        mst_cost += w;
        process(u); // take u, process all edges incident to u
      }
    }                                        // each edge is in pq only once!
    System.out.printf("MST cost = %d (Prim's)\n", mst_cost);
  }
}

File: ch4/kruskal_prim.cpp
// February 2019 note:
// This code uses new C++17 structured binding
// use this compiler setting "g++ -g -O2 -static -std=gnu++17 {files}"

#include <bits/stdc++.h>
using namespace std;

typedef tuple<int, int, int> iii;
typedef pair<int, int> ii;
typedef vector<int> vi;
typedef vector<ii> vii;

// Union-Find Disjoint Sets Library written in OOP manner
// using both path compression and union by rank heuristics
class UnionFind {                                // OOP style
private:
  vi p, rank, setSize;                           // vi is vector<int>
  int numSets;
public:
  UnionFind(int N) {
    p.assign(N, 0); for (int i = 0; i < N; i++) p[i] = i;
    rank.assign(N, 0);
    setSize.assign(N, 1);
    numSets = N;
  }
  int findSet(int i) { return (p[i] == i) ? i : (p[i] = findSet(p[i])); }
  bool isSameSet(int i, int j) { return findSet(i) == findSet(j); }
  void unionSet(int i, int j) {
    if (!isSameSet(i, j)) {
      numSets--;
      int x = findSet(i), y = findSet(j);
      // rank is used to keep the tree short
      if (rank[x] > rank[y]) {
        p[y] = x;
        setSize[x] += setSize[y];
      }
      else {
        p[x] = y;
        setSize[y] += setSize[x];
        if (rank[x] == rank[y]) rank[y]++;
      }
    }
  }
  int numDisjointSets() { return numSets; }
  int sizeOfSet(int i) { return setSize[findSet(i)]; }
};

vector<vii> AL;                                  // the graph stored in AL
vi taken;                                        // to avoid cycle
priority_queue<ii> pq;                           // to select shorter edges
// note: the default setting for C++ STL priority_queue is a max heap
// so we use -ve sign to reverse the sort order

void process(int u) { // set u as taken and enqueue neighbors of u
  taken[u] = 1;
  for (auto &v : AL[u])
    if (!taken[v.first])
      pq.push({-v.second, -v.first});            // sort by (inc) weight
}                                                // then by (inc) id

int main() {
  /*
  // Graph in Figure 4.10 left, format: list of weighted edges
  // This example shows another form of reading graph input
  5 7
  0 1 4
  0 2 4
  0 3 6
  0 4 6
  1 2 2
  2 3 8
  3 4 9
  */

  freopen("kruskal_prim_in.txt", "r", stdin);

  int V, E; scanf("%d %d", &V, &E);
  // Kruskal's algorithm merged with Prim's algorithm
  AL.assign(V, vii());
  vector<iii> EL(E);
  for (int i = 0; i < E; i++) {
    int u, v, w; scanf("%d %d %d", &u, &v, &w);  // read as (u, v, w)
    EL[i] = tie(w, u, v);                        // reorder as (w, u, v)
    AL[u].emplace_back(v, w);
    AL[v].emplace_back(u, w);
  }
  sort(EL.begin(), EL.end());                    // sort by w, O(E log E)
  // note: std::tuple has built-in comparison function

  int mst_cost = 0, num_taken = 0;
  UnionFind UF(V);                               // all V are disjoint sets
  for (int i = 0; i < E && num_taken < V-1; i++) { // for each edge, O(E)
    // int w, u, v; tie(w, u, v) = EL[i];        // C++11 style
    auto [w, u, v] = EL[i];                      // C++17 style
    if (!UF.isSameSet(u, v)) {                   // check
      num_taken++;                               // 1 more edge is taken
      mst_cost += w;                             // add w of this edge
      UF.unionSet(u, v);                         // link them
    } // note: the runtime cost of UFDS is very light
  }
  // note: the number of disjoint sets must eventually be 1 for a valid MST
  printf("MST cost = %d (Kruskal's)\n", mst_cost);

// inside int main() --- assume the graph is stored in AL, pq is empty
  taken.assign(V, 0);                            // no vertex is taken
  process(0);                                    // take+process vertex 0
  mst_cost = 0; num_taken = 0;                   // no edge has been taken
  while (!pq.empty() && num_taken < V-1) {       // until we take V-1 edges
    // int w, u; tie(w, u) = pq.top(); pq.pop(); // C++11 style
    auto [w, u] = pq.top(); pq.pop();            // C++17 style
    w = -w; u = -u;                              // negate to reverse order
    if (!taken[u]) {                             // we have not take u yet
      num_taken++;                               // 1 more edge is taken
      mst_cost += w;                             // add w of this edge
      process(u);                                // take+process vertex u
    }                                            // each edge is in pq once
  }
  printf("MST cost = %d (Prim's)\n", mst_cost);

  return 0;
}
